User Preferences & QOTW Reminder

DynxstyGIT opened the pull request:
This PR introduces the `/preferences` command which lets users easily control the state of some predefined preferences.
As this command cannot come alone, this also introduces the QOTW Reminder @sahilasopa mentioned some time ago.

danthe1st commented on src/main/java/net/javadiscord/javabot/systems/user_preferences/model/UserPreference.java:
```diff
@@ -0,0 +1,17 @@
+@Data
+public class UserPreference {
+	private long userId;
+	private int ordinal;
```
Maybe use the `Preference` enum here (if this is possible with the repository)

DynxstyGIT replied:
yeah, but just saving the ordinal takes up less space instead of saving it as a string

danthe1st replied:
I didn't men saving the string in the DB. I meant using the `Preference` object here and still saving the ordinal to the DB.

DynxstyGIT replied:
Oh, so you mean just merging the enum with the model class. Will do that!

DynxstyGIT replied:
![image](https://user-images.githubusercontent.com/48297101/182050039-926e17d4-29ce-4450-98c1-5b7275c2c847.png)
sike

danthe1st replied:
No, I meant having the enum as an attribute of the model class and just saving the ordinal value in the DB.

DynxstyGIT replied:
oh yeah

DynxstyGIT commented on src/main/java/net/javadiscord/javabot/Bot.java:
```diff
@@ -109,7 +109,7 @@ public static void main(String[] args) throws Exception {
 		dih4jda = DIH4JDABuilder.setJDA(jda)
 				.setCommandsPackage("net.javadiscord.javabot")
-				.setDefaultCommandType(RegistrationType.GUILD)
+				.setDefaultCommandType(RegistrationType.GLOBAL)
```
Yup, DIH4JDA automatically does that

DynxstyGIT commented on src/main/java/net/javadiscord/javabot/systems/qotw/submissions/SubmissionManager.java:
```diff
@@ -129,6 +130,21 @@ public boolean hasActiveSubmissionThreads(long authorId) {
+			return repo.getSubmissionByQuestionNumber(repo.getCurrentQuestionNumber());
+		} catch (SQLException e) {
+			ExceptionLogger.capture(e, getClass().getSimpleName());
+			return List.of();
```
Well, yeah. If it can't retrieve the submissions then yeah, the list's empty.

DynxstyGIT commented on src/main/java/net/javadiscord/javabot/systems/user_preferences/commands/UserPreferenceCommand.java:
```diff
@@ -0,0 +1,67 @@
+	 */
+	public UserPreferenceCommand() {
+		setSlashCommandData(Commands.slash("preferences", "Allows you to set some preferences!")
+				.addOption(OptionType.INTEGER, "preference", "The preference to set.", true, true)
```
no, this would make the whole thing not as dynamic as it is. the autocomplete shows all preferences including the users' current state 
![image](https://user-images.githubusercontent.com/48297101/182049214-9c8584e3-93b7-4d95-8653-0dfe389e0043.png)

danthe1st replied:
Note that this comes at the cost that autocompletes are not responsive. It may take multiple seconds for the autocomplete to appear (depending on latency, device type, etc.).

DynxstyGIT replied:
Well. yeah, that might be possible. We'll see how this goes ig

DynxstyGIT commented:
I've now replaced the autocomplete with a "dynamic" choice option. 
![image](https://user-images.githubusercontent.com/48297101/182118512-68ec6509-9b3b-477c-8888-ddd0c20b8491.png)

In addition to that, i've added a `/preferences list` command:
![image](https://user-images.githubusercontent.com/48297101/182118627-9b6ed383-09e3-478c-8625-6664479b05f7.png)

danthe1st commented on src/main/java/net/javadiscord/javabot/systems/user_preferences/commands/PreferencesListSubcommand.java:
```diff
@@ -0,0 +1,35 @@
+		String preferences = Arrays.stream(Preference.values())
+				.map(p -> String.format("`%s` %s", manager.getOrCreate(event.getUser().getIdLong(), p).isEnabled() ? "\uD83D\uDFE2" : "\uD83D\uDD34", p))
+				.collect(Collectors.joining("\n"));
+		Responses.info(event, String.format("%s's Preferences", event.getUser().getName()), preferences).queue();
```
You might want to use ephemeral responses here.

DynxstyGIT replied:
As seen in the image above, the response is already ephemeral

danthe1st commented on src/main/java/net/javadiscord/javabot/systems/user_preferences/model/Preference.java:
```diff
@@ -1,22 +1,34 @@
+	}
+
+	public Command.Choice toChoice() {
+		return new Command.Choice(name, String.valueOf(ordinal()));
```
I don't think `Preference` should have JDA-specific code. Instead, you should move that to the subcommand (or a utility class).

DynxstyGIT replied:
yup
